cloud.foundry.cli/src/test/java/cloud/foundry/cli/logic/ApplyLogicTest.java

```diff
@@ -97,4 +100,57 @@ public void testApplyApplicationsCreatesApplication() {
         assertThat(manifest.getBuildpack(), is("someBuildpack"));
     }
 
+    @Test
+    public void testApplyServicesWithNull() {
+        ApplyLogic applyLogic = new ApplyLogic(mock(DefaultCloudFoundryOperations.class));
+
+        assertThrows(NullPointerException.class, () -> applyLogic.applyServices(null));
+    }
+
+    @Test
+    public void testApplyServices() {
+        //given
+        DefaultCloudFoundryOperations cfOperationsMock = mock(DefaultCloudFoundryOperations.class);
+
+        //desired Services
+        HashMap<String, ServiceBean> desiredServices = new HashMap<>();
+        ServiceBean serviceBean = new ServiceBean();
+        desiredServices.put("exampleService", serviceBean);
+
+        //liveConfig
+        HashMap<String, ServiceBean> liveConfig = new HashMap<>();
+        GetLogic getlogic = mock(GetLogic.class);
+        when(getlogic.getServices(any())).thenReturn(liveConfig);
+
+        //Diffresult
+        DiffLogic diffLogicMock = mock(DiffLogic.class);
+        DiffResult diffResultMock = mock(DiffResult.class);
+        when(diffLogicMock.createDiffResult(any(), any())).thenReturn(diffResultMock);
+
+        //allServiceChanges
+        List<CfChange> cfChanges = new LinkedList<>();
+        CfChange cfChangeMock = mock(CfChange.class);
+        cfChanges.add(cfChangeMock);
+        Map<String, List<CfChange>> allServiceChanges = mock(Map.class);
+        when(diffResultMock.getServiceChanges()).thenReturn(allServiceChanges);
+        //entry set of allServiceChanges
+        Set<Map.Entry<String, List<CfChange>>> entrySet = new HashSet<>();
+        Map.Entry entry = mock(Map.Entry.class);
+        entrySet.add(entry);
+        when(allServiceChanges.entrySet()).thenReturn(entrySet);
+
+
+        //when
+        ApplyLogic applyLogic = new ApplyLogic(cfOperationsMock);
+        applyLogic.setDiffLogic(diffLogicMock);
```

Comment:
As said above, you should consider using DI instead of these setters. Why not add a proper constructor which requires both a custom diff and get logic?

Reply:
Setters also work as DI. Because the diff and get logic has no state, I can create it for the user (no need for multiple constructors/ multiple constructor arguments). Thta is why I prefer the DI over the setters in this sitaution.